### PR TITLE
update(grid): Add smart grid declarations, add media-query to mixins, add sass maps to global variables

### LINKED ISSUE #480: Grid system enhancements
Add css support for column definitions on grid container. This would eliminate the need to set a size on each grid item, while maintaining the ability to override size on any grid item. 
* Add sass maps to variables.scss
* Create column definitions for grid container, including responsive definitions `.pf-l-grid.pf-m-all-{1-12}-col`

Update grid GridMediaQuery() mixin in src/patternfly/layouts/Grid/styles.scss.

* change this:
```
@else if $point == "sm" or $point == "" or $point == null {
  @media screen and (min-width: $pf-global--breakpoint--sm) {
    @content;
  }
}
```
* to this:
```
@else if $point == "sm" {
  @media screen and (min-width: $pf-global--breakpoint--sm) {
    @content;
  }
}
```
* and add this
```
@if $point == "" or $point == null {
  @content;
}
```

### PR DESCRIPTION
**Pull request by mattnolting**
- Adds smart grid view
- Updates documentation
- Updates partials

Closes #480

### CODE REVIEW

File: src/patternfly/layouts/Grid/docs/code.md [hunk 1]
```diff
@@ -1,14 +1,14 @@
 ## Overview
 
-A Grid layout framework including responsive columns, row spans and gutters 
+A Grid layout framework including responsive columns, row spans and gutters. **Breakpoint is optional. Breakpoints: -on-xs, -on-sm, -on-md, -on-lg, -on-xl**
```
Comment (matthewcarleton) [inline src/patternfly/layouts/Grid/docs/code.md]: Should these classes be in the usage table as well?

File: src/patternfly/layouts/Grid/docs/code.md [hunk 2]
```diff
@@ -1,14 +1,14 @@
...
 
 | Class | Applied To | Outcome |
 | -- | -- | -- |
 | `.pf-l-grid` | `<div>` | Initializes the grid layout. |
 | `.pf-l-grid__item` | `<div>` | Explicitly sets a child of the grid. This class isn't necessary, but it is included to keep inline with BEM convention, and to provide an entity that will later be used for applying modifiers. |
-| `.pf-m-gutters` | `.pf-l-grid` | Adds space between children |
-| `.pf-m-{1-12}-col` | `.pf-l-grid__item` | Defines grid item width.  Although not required, they are strongly suggested. If not used, grid item will default to 12 col. |
-| `.pf-m-{1-12}-col-on-{breakpoint}` | `.pf-l-grid__item` | Defines grid-item width based on breakpoint.  Breakpoints: xs, sm, md, lg, xl |
-| `.pf-m-{2-x}-row`, `.pf-m-{1-11}-col` | `.pf-l-grid__item` | Defines grid item row span.  For row spans to function correctly, the value of of the current row plus the grid items to span must be equal to or less than 12. Example: .pf-m-8-col.pf-m-2-row + .pf-m-4-col + .pf-m-4-col. There is no limit to number of spanned rows. |
+| `.pf-m-gutters` | `.pf-l-grid` | Adds grid-gap. |
```
Comment (matthewcarleton) [inline src/patternfly/layouts/Grid/docs/code.md]: is it strange to have the modifier say gutter when the property is called gap? Maybe it doesn't matter.

File: src/patternfly/layouts/Grid/examples/grid-base-example.hbs [hunk 3]
```diff
@@ -1,35 +1,35 @@
 {{#> grid}}
-  {{#> grid-item modifierClass="pf-m-12-col"}}
+  {{#> grid-item grid-item--modifiers="pf-m-12-col"}}
```
Comment (matthewcarleton) [inline src/patternfly/layouts/Grid/examples/grid-base-example.hbs]: This should be grid-item--modifier

File: src/patternfly/layouts/Grid/grid.hbs [hunk 4]
```diff
@@ -1,3 +1,6 @@
-<div class="pf-l-grid {{ modifierClass }}" >
+<div class="pf-l-grid{{#if grid--modifiers}} {{grid--modifiers}}{{/if}}"
```
Comment (matthewcarleton) [inline src/patternfly/layouts/Grid/grid.hbs]: this will have to change when the "s" is dropped from modifiers.

**Comment by dmiller9911** [comment]
Should it be possible to have an item that has its own size?  IE. I have a grid that has 6-col for all, but I want 1 to have 12-col?  Right now that seems overwritten from specificity.

File: src/patternfly/layouts/Grid/docs/code.md (hunk 2 above)
Comment (mattnolting) [reply src/patternfly/layouts/Grid/docs/code.md]: I agree, probably should change to `.pf-m-grid-gap` and add `.pf-m-column-gap` and `.pf-m-row-gap`

File: src/patternfly/layouts/Grid/examples/grid-base-example.hbs (hunk 3 above)
Comment (mattnolting) [reply src/patternfly/layouts/Grid/examples/grid-base-example.hbs]: Changed this to plural form to match `grid-item--attributes` as multiple modifiers can be passed to partials.

Comment (matthewcarleton) [reply src/patternfly/layouts/Grid/examples/grid-base-example.hbs]: ah gotcha, makes sense.

File: src/patternfly/layouts/Grid/docs/code.md (hunk 2 above)
Comment (andresgalante) [reply src/patternfly/layouts/Grid/docs/code.md]: Even though we know the property is a grid gap the thing that generates is a gutter. This is also shared with the different layouts that don't use grid. For example flex and margins on split. They all share the same modifier: gutter

File: src/patternfly/layouts/Grid/docs/code.md (hunk 1 above)
Comment (mattnolting) [reply src/patternfly/layouts/Grid/docs/code.md]: Updated the usage doc to better show how to form responsive classes

**Comment by mattnolting** [comment]
@andresgalante Updated `pf-m-gutter` to `pf-m-gap`
@dmiller9911 Updated css to honor overrides

- Made a few changes to the examples, including a new section demonstrating smart grid overrides. 
- Updated the documentation to better reflect responsive selector construction

**Comment by andresgalante** [comment]
Please do not change `pf-m-gutter` to `pf-m-gap`.

Maybe I wasn't clear on my comment, `pf-m-gap` is wrong since we can't reuse it on other components that are using flex.

`pf-m-gutter` is correct.

**Comment by mattnolting** [comment]
@andresgalante reverted

**Comment by andresgalante** [comment]
@srambach sorry, I just saw your comment about the modifier ve modifiers. @mattnolting would you mind updating it?

**Comment by mattnolting** [comment]
@andresgalante @srambach I understand. How do you feel about me creating a few housekeeping issues regarding standardization on this? I strongly believe this is the correct usage, and current guidelines and template generation need to be updated.. Also, the param construction varies quite a bit throughout the components already. 
Examples: 
`pf-c-alert--Modifier`
`pf-c-badge--modifier`
`btnClass`
`<div class="pf-c-content {{ modifierClass }}">`
`pf-c-dropdown--Modifier` 
`modal-box--modifiers`

The correct construction should be: `[partial]--modifiers` + `[partial]--attributes`

For example, in: `component.hbs`
```
<div class="classname{{#if component--modifiers}} {{component--modifiers}}{{/if}}"
  {{#if component--attributes}}
    {{{component--attributes}}}
  {{/if}}
</div>

component--modifiers relates to the partial itself, rather than the assigned class

```
Case for:
- `item-header--` rather than `pf-c-item-header--`
Classnames can change. Using partial name (instead of classname) means that if a selector changes, you won't have to change every instance where that partial is called and modified (in example files).

- `--modifiers` rather than `--modifier`
Multiple modifiers can be applied to an element. This is grammatically correct and matches `--attributes`, as multiple attributes can be applied.

- Bear in mind that `component--modifiers` is already scoped to that `component`. So it's unnecessary to namespace with `pf-c-`. If passing multiple values to a partial, prefixing them makes less scannable and gets repetitive.

```
{{#> partial partial--modifiers="special" partial--attributes='id="[id]" aria-label="[label]" aria-describedby="[description]"' expandable="true" expanded="true"}}
  content
{{/partial}}
```

rather than

```
{{#> partial pf-c-partial--modifiers="special" pf-c-partial--attributes='id="[id]" aria-label="[label]" aria-describedby="[description]"'  pf-c-partial--expandable="true" pf-c-partial--expanded="true"}}
  content
{{/partial}}
```
What do you think?

**Pull request merged**